src/include/duckdb/execution/reservoir_sample.hpp

```diff
@@ -182,10 +237,115 @@ class ReservoirSamplePercentage : public BlockingSample {
 
 	//! The set of finished samples of the reservoir sample
 	vector<unique_ptr<ReservoirSample>> finished_samples;
+
 	//! The amount of tuples that have been processed so far (not put in the reservoir, just processed)
 	idx_t current_count = 0;
-	//! Whether or not the stream is finalized. The stream is automatically finalized on the first call to GetChunk();
+	//! Whether or not the stream is finalized. The stream is automatically finalized on the first call to
+	//! GetChunkAndShrink();
 	bool is_finalized;
 };
 
+struct SampleCopyHelper {
+	SelectionVector sel;
+	//! Priority queue of [random element, index] for each of the elements in the sample
+	std::priority_queue<std::pair<double, idx_t>> reservoir_weights;
+	vector<idx_t> actual_indexes;
+};
+
+class IngestionSample : public BlockingSample {
+public:
+	static constexpr const SampleType TYPE = SampleType::INGESTION_SAMPLE;
+
+	constexpr static idx_t FIXED_SAMPLE_SIZE_MULTIPLIER = 10;
+	constexpr static idx_t FAST_TO_SLOW_THRESHOLD = 60;
+
+	// how much of every chunk we consider for sampling.
+	// This can be lowered later to improve ingestion performance.
+	constexpr static double CHUNK_SAMPLE_PERCENTAGE = 1.00;
+	// If the table has less than 204800 rows, this is the percentage
+	// of values we save when serializing/returning a sample.
+	constexpr static double SAVE_PERCENTAGE = 0.01;
+
+	IngestionSample(Allocator &allocator, int64_t seed);
+	explicit IngestionSample(idx_t sample_count, int64_t seed = 1);
+
+	unique_ptr<BlockingSample> ConvertToReservoirSample();
```

Comment:
It would be great if the `IngestionSample` would inherit from the `ReservoirSample` directly - or at least if more code/functionality was shared. For example, the appending code with the larger data chunks and the selection vector could be moved to the (base) reservoir sample class.